### PR TITLE
Async sampling

### PR DESCRIPTION
**Pull request by lucasavila00**
```
./target/profiling/mistralrs-bench -p 0 -g 64 -r 1 -c 8  gguf -t mistralai/Mistral-7B-Instruct-v0.1 -m TheBloke/Mistral-7B-Instruct-v0.1-GGUF -f mistral-7b-instruct-v0.1.Q4_K_M.gguf
```

Master

image

This PR

image

### CODE REVIEW

File: mistralrs-core/src/pipeline/mod.rs [hunk 1]
```diff
@@ -319,59 +320,62 @@ pub trait Pipeline: Send + Sync {
...
+    let start_at = seq.get_toks().len().saturating_sub(repeat_last_n);
+
+    let sampler = seq.sampler();
+    let logits_clone = logits.clone();
+    let ctx_clone = seq.get_toks()[start_at..].to_vec();
+    let rng_clone = rng.clone();
+    let first_lobprobs_response = tokio_rayon::spawn(move || {
+        sampler.sample(logits_clone, Some(&ctx_clone), return_logprobs, rng_clone)
+    })
+    .await?;
```
Comment (lucasavila00) [inline mistralrs-core/src/pipeline/mod.rs:344]: run this in a threadpool, but using async and await for control flow

**Comment by EricLBuehler** [comment]
@lucasavila00, thanks for your work here. I think we definitely need to pursue the throughput decrease as the batch increase. I like the idea of async being incorporated, did this not work out?

**Comment by lucasavila00** [comment]
@EricLBuehler It had mixed results.

On bs=8 it improved from 5ms to 2ms.

But on bs=1 it got worse, from 0.8ms to 1.2ms.

Clippy raised an issue of a non-async mutex being held across an await point.

The best fix for it would be to use an async aware mutex? Or drop the lock a lot of times and re-lock when needed?

I'm still learning async rust and don't feel confident enough to work on this MR yet, as it requires defining the overall async structure for the engine.

Also, profiling CPU code became harder (eg: samply profiler). Only the nvidia profiler showed interpretable results.

**Comment by EricLBuehler** [comment]
> On bs=8 it improved from 5ms to 2ms.
Great! Perhaps we could profile it and see when the performance gains go away, and use this then.
> Clippy raised an issue of a non-async mutex being held across an await point.
We could use [this type](https://doc.servo.org/tokio/sync/struct.Mutex.html).

I think this sort of structure is very interesting, I'll take a look in the next few days. Thanks for working on it.

**Comment by lucasavila00** [comment]
I'll leave it open, so it picks up my commits.

I just fixed the clippy issue, by not holding the lock across awaits.

**Comment by lucasavila00** [comment]
@EricLBuehler cargo test fails because it can't download the mistral tokenizer from HF.

Any chance the CI account does not have access to the model, which has been recently locked (locked like llama where one needs to request access)

**Comment by lucasavila00** [comment]
> Great! Perhaps we could profile it and see when the performance gains go away, and use this then.

I implemented it in the last commit. It only uses the async pool if there's more than one batch. Then, the performance loss is gone.

It means that the async code is free if not used.

**Comment by lucasavila00** [comment]
I added 2 nvidia profiler screenshots to the main PR body.

**Comment by EricLBuehler** [comment]
@lucasavila00 

> cargo test fails because it can't download the mistral tokenizer from HF.

I set the `HF_TOKEN` secret during CI: 
> https://github.com/EricLBuehler/mistral.rs/blob/9ec90a40949efe9c68828c048da136874d1ba2f1/.github/workflows/ci.yml#L49

**Comment by EricLBuehler** [comment]
@lucasavila00, it looks like there are some merge conflicts.

**Comment by lucasavila00** [comment]
> @lucasavila00, it looks like there are some merge conflicts.

I'm fixing it

File: mistralrs-core/src/engine/mod.rs [hunk 2]
```diff
@@ -213,21 +243,113 @@ impl Engine {
...
+        pipeline: Arc<Mutex<dyn Pipeline>>,
         seqs: &mut [&mut Sequence],
         logits: Tensor,
         prefix_cacher: &mut PrefixCacheManager,
         disable_eos_stop: bool,
+        rng: Arc<Mutex<Isaac64Rng>>,
     ) -> Result<()> {
         let seqs_len = seqs.len();
-        let logits_seq = logits.chunk(seqs_len, 0).unwrap();
+        let logits_seq = logits.to_device(&Device::Cpu)?.chunk(seqs_len, 0)?;
```
Comment (lucasavila00) [inline mistralrs-core/src/engine/mod.rs:321]: Now that we do a synchronization before we start sampling, we can get statistics about sampling speed.

Basically reverting https://github.com/EricLBuehler/mistral.rs/pull/151

**Comment by lucasavila00** [comment]
Master

```
+------------------------------------+---------+--------+--------------+--------------+-------------+--------------+
| model                              | backend | test   | t/s          | ms/t         | concurrency | throughput/s |
+------------------------------------+---------+--------+--------------+--------------+-------------+--------------+
| mistralai/Mistral-7B-Instruct-v0.1 | CUDA    | tg 128 | 58.712±0.644 | 17.034±0.190 |           1 |    58.712006 |
+------------------------------------+---------+--------+--------------+--------------+-------------+--------------+
+------------------------------------+---------+--------+--------------+--------------+-------------+--------------+
| model                              | backend | test   | t/s          | ms/t         | concurrency | throughput/s |
+------------------------------------+---------+--------+--------------+--------------+-------------+--------------+
| mistralai/Mistral-7B-Instruct-v0.1 | CUDA    | tg 128 | 45.956±0.805 | 21.766±0.380 |           2 |      91.9128 |
+------------------------------------+---------+--------+--------------+--------------+-------------+--------------+
+------------------------------------+---------+--------+--------------+--------------+-------------+--------------+
| model                              | backend | test   | t/s          | ms/t         | concurrency | throughput/s |
+------------------------------------+---------+--------+--------------+--------------+-------------+--------------+
| mistralai/Mistral-7B-Instruct-v0.1 | CUDA    | tg 128 | 29.011±0.321 | 34.473±0.388 |           4 |     116.0458 |
+------------------------------------+---------+--------+--------------+--------------+-------------+--------------+
+------------------------------------+---------+--------+--------------+--------------+-------------+--------------+
| model                              | backend | test   | t/s          | ms/t         | concurrency | throughput/s |
+------------------------------------+---------+--------+--------------+--------------+-------------+--------------+
| mistralai/Mistral-7B-Instruct-v0.1 | CUDA    | tg 128 | 15.469±0.388 | 64.685±1.639 |           8 |    123.75505 |
+------------------------------------+---------+--------+--------------+--------------+-------------+--------------+
```


This

```
+------------------------------------+---------+--------+--------------+--------------+-------------+--------------+
| model                              | backend | test   | t/s          | ms/t         | concurrency | throughput/s |
+------------------------------------+---------+--------+--------------+--------------+-------------+--------------+
| mistralai/Mistral-7B-Instruct-v0.1 | CUDA    | tg 128 | 58.752±0.501 | 17.022±0.147 |           1 |    58.752266 |
+------------------------------------+---------+--------+--------------+--------------+-------------+--------------+
+------------------------------------+---------+--------+--------------+--------------+-------------+--------------+
| model                              | backend | test   | t/s          | ms/t         | concurrency | throughput/s |
+------------------------------------+---------+--------+--------------+--------------+-------------+--------------+
| mistralai/Mistral-7B-Instruct-v0.1 | CUDA    | tg 128 | 48.121±0.124 | 20.781±0.053 |           2 |     96.24125 |
+------------------------------------+---------+--------+--------------+--------------+-------------+--------------+
+------------------------------------+---------+--------+--------------+--------------+-------------+--------------+
| model                              | backend | test   | t/s          | ms/t         | concurrency | throughput/s |
+------------------------------------+---------+--------+--------------+--------------+-------------+--------------+
| mistralai/Mistral-7B-Instruct-v0.1 | CUDA    | tg 128 | 30.030±0.020 | 33.300±0.022 |           4 |    120.12018 |
+------------------------------------+---------+--------+--------------+--------------+-------------+--------------+
+------------------------------------+---------+--------+--------------+--------------+-------------+--------------+
| model                              | backend | test   | t/s          | ms/t         | concurrency | throughput/s |
+------------------------------------+---------+--------+--------------+--------------+-------------+--------------+
| mistralai/Mistral-7B-Instruct-v0.1 | CUDA    | tg 128 | 16.839±0.008 | 59.384±0.027 |           8 |    134.71562 |
+------------------------------------+---------+--------+--------------+--------------+-------------+--------------+
```

**Comment by EricLBuehler** [comment]
@lucasavila00, this looks good. However, I think there is one more merge conflict.

File: mistralrs-core/src/pipeline/mod.rs [hunk 3]
```diff
@@ -5,6 +5,7 @@ mod loaders;
 mod macros;
 mod normal;
 use crate::aici::toktree::TokTrie;
+use crate::DeviceMapMetadata;
```
Comment (EricLBuehler) [inline mistralrs-core/src/pipeline/mod.rs:8]: We should remove this, it is causing the error.
```suggestion

```

File: mistralrs-core/src/pipeline/mod.rs [hunk 4]
```diff
@@ -5,6 +5,7 @@ mod loaders;
...
 use crate::{api_dir_list, api_get_file, DeviceMapMetadata};
 use crate::{get_bias_if_not_allowed, sampler::Logprobs, sequence::SequenceRecognizer};
```
Comment (EricLBuehler) [inline mistralrs-core/src/pipeline/mod.rs:10]: We should remove this line, the imports are unused.
```suggestion
```

**Review by EricLBuehler — CHANGES_REQUESTED** [review]
This looks good. I think I made some mistakes when doing the conflict resolution, so I've marked them.

**Review by EricLBuehler — APPROVED** [review]
Looks good, thank you for adding this!

**Comment by EricLBuehler** [comment]
@lucasavila00, I think there are unfortunately still some merge conflicts.

**Comment by lucasavila00** [comment]
@EricLBuehler please consider squashing this MR

Or please let me know if I should squash it

**Review by EricLBuehler — APPROVED** [review]

**Comment by EricLBuehler** [comment]
@lucasavila00 thank you for adding this!
